src/internal/routing-table-getter/routing-table-factory.js

```diff
@@ -0,0 +1,210 @@
+ *
+ * @param {ServerAddress} routerAddress The router address, it is used for loggin purposes
+ * @param {function():Promise} runSupplier The supplier which will be runned
+ * @return {Promise} the result of the supplier
+ */
+export async function runWithExceptionTreament (routerAddress, runSupplier) {
+  try {
+    return await runSupplier()
+  } catch (error) {
+    if (error.code === DATABASE_NOT_FOUND_CODE) {
+      throw error
+    } else if (error.code === PROCEDURE_NOT_FOUND_CODE) {
+      // throw when getServers procedure not found because this is clearly a configuration issue
+      throw newError(
+        `Server at ${routerAddress.asHostPort()} can't perform routing. Make sure you are connecting to a causal cluster`,
+        SERVICE_UNAVAILABLE
+      )
+    } else {
+      // return nothing when failed to connect because code higher in the callstack is still able to retry with a
+      // different session towards a different router
+      return null
+    }
+  }
+}
+
+/**
+ * Create a valid routing table from a raw object
+ *
+ * @param {string} database the database name. It is used for logging purposes
+ * @param {ServerAddress} routerAddress The router address, it is used for loggin purposes
+ * @param {RawRoutingTable} rawRoutingTable Method used to get the raw routing table to be processed
+ * @param {RoutingTable} The valid Routing Table
+ */
+export function createValidRoutingTable (
+  database,
+  routerAddress,
+  rawRoutingTable
+) {
+  const expirationTime = parseTtl(rawRoutingTable, routerAddress)
+  const { routers, readers, writers } = parseServers(
+    rawRoutingTable,
+    routerAddress
+  )
+
+  assertNonEmpty(routers, 'routers', routerAddress)
+  assertNonEmpty(readers, 'readers', routerAddress)
+
+  return new RoutingTable({
+    database,
+    routers,
+    readers,
+    writers,
+    expirationTime
+  })
+}
+
+/**
+ * Parse server from the RawRoutingTable.
+ *
+ * @param {RawRoutingTable} rawRoutingTable the raw routing table
+ * @param {string} routerAddress the router address
+ * @returns {Object} The object with the list of routers, readers and writers
+ */
+function parseServers (rawRoutingTable, routerAddress) {
+  try {
+    let routers = []
+    let readers = []
+    let writers = []
+
+    rawRoutingTable.servers.forEach(server => {
+      const role = server.role
+      const addresses = server.addresses
+
+      if (role === 'ROUTE') {
+        routers = parseArray(addresses).map(address =>
+          ServerAddress.fromUrl(address)
+        )
+      } else if (role === 'WRITE') {
+        writers = parseArray(addresses).map(address =>
+          ServerAddress.fromUrl(address)
+        )
+      } else if (role === 'READ') {
+        readers = parseArray(addresses).map(address =>
+          ServerAddress.fromUrl(address)
+        )
+      } else {
+        throw newError('Unknown server role "' + role + '"', PROTOCOL_ERROR)
```

Comment:
I would prefer to not consider unknowns an error, as long as we get the roles that we want it is probably fine to go on

Reply:
It's a good point, i will skip if it's unknown